tests/unit/generator_test.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2024 ScyllaDB Ltd.
+ */
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/async_generator.hh>
+#include <seastar/coroutine/exception.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/util/bool_class.hh>
+#include <seastar/util/later.hh>
+
+using namespace seastar;
+
+using do_suspend = bool_class<struct do_suspend_tag>;
+
+namespace {
+
+struct always_yield_awaiter final {
+    bool await_ready() const {
+        return false;
+    }
+
+    template <typename T>
+    void await_suspend(std::coroutine_handle<T> h) {
+        schedule(&h.promise());
+    }
+
+    void await_resume() {
+    }
+};
+
+}
+
+coroutine::experimental::async_generator<int>
+fibonacci_sequence(unsigned count,
+                   do_suspend suspend) {
+    auto a = 0, b = 1;
+    for (unsigned i = 0; i < count; ++i) {
+        if (std::numeric_limits<decltype(a)>::max() - a < b) {
+            throw std::out_of_range(
+                fmt::format("fibonacci[{}] is greater than the largest value of int", i));
+        }
+        if (suspend) {
+            co_await yield();
+        }
+        co_yield std::exchange(a, std::exchange(b, a + b));
+    }
+}
+
+seastar::future<> test_async_generator_drained(do_suspend suspend) {
+    auto expected_fibs = {0, 1, 1, 2};
+    auto expected_fib = std::begin(expected_fibs);
+
+    auto actual_fibs = fibonacci_sequence(std::size(expected_fibs), suspend);
```

Comment:
Can you please add test cases for move and swap of the async_generator?

Reply:
sure. added.